docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+        - Description: ProviderSpec details Provider-specific configuration to use during node creation.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `MachineVersionInfo`
+        - Description: Versions of key software to use. This field is optional at cluster creation time, and omitting the field indicates that the cluster installation tool should select defaults for the user. These defaults may differ based on the cluster installer, but the tool should populate the values it uses when persisting Machine objects. A Machine spec missing this field at runtime is invalid.
+- **To add**
+    - **Bootstrap**
+        - Type: `Bootstrap`
+        - Description: Bootstrap is a reference to a local struct which encapsulates fields to configure the Machine’s bootstrapping mechanism.
+    - **InfrastructureRef [required]**
+        - Type: `corev1.TypedLocalObjectReference`
+        - Description: InfrastructureRef is a required reference to a Custom Resource Definition offered by an infrastructure provider.
+    - **Version [optional]**
+        - Type: `string`
+        - Description: Version defines the desired Kubernetes version. This field is meant to be optionally used by bootstrap providers.
+
+```go
+type MachineStatus struct
+```
+
+- **To remove**
+    - **ProviderStatus [optional]**
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderStatus details a Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `*MachineVersionInfo`
+        - Description: Versions specifies the current versions of software on the corresponding Node (if it exists). This is provided for a few reasons [...].
+    - **Phase [optional]**  _Superseded by Phase_
+        - Type: `*string`
+        - Description: Phase represents the current phase of machine actuation. E.g. Pending, Running, Terminating, Failed etc.
+- **To add**
+    - **Phase**
+        - Type: `MachinePhaseType`
+        - Description: Phase represents the current state of a Machine. This field is not meant to be used as a source of truth. Controllers working on Machines should always respect the documented transition conditions to determine the state.
+    - **Version [optional]**
+        - Type: ``*string`
+        - Description: Version specifies the current version of Kubernetes running on the corresponding Node. This is meant to be a means of bubbling up status from the Node to the Machine. It is entirely optional, but useful for end-user UX if it’s present.
+    - **BootstrapPhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **InfrastructurePhase [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
```

Comment:
Isn't the Bootstrap.ConfigRef is going to point to a provider-specific CR like you describe?

Reply:
If so then it needs a better name. When I saw `ConfigRef` I assumed it pointed to a `ConfigMap`.

Reply:
I agree. ProviderDataRef?

Reply:
I do like the idea of finding a better name, but I would like to avoid `ProviderDataRef`, since I think it could cause confusion based on having only a singular "Provider" in v1alpha1.

Reply:
Hopeful clarification: `machine.spec.bootstrap` is the portion of the machine spec related to bootstrapping. `machine.spec.bootstrap.configRef` (name pending) is a reference to a custom resource that holds bootstrap configuration details for this machine. We want to use a custom resource for bootstrap configuration details so it can fully participate in validating webhooks.

Reply:
Yeah, I thought since the reference was inside the Bootstrap struct there wouldn't be any confusion about ProviderDataRef referring to the bootstrap provider. Maybe not, though.

I definitely think we want it to be an object reference. I agree that the name ConfigRef may be confusing as to what is expected there.

For the infrastructure parts we don't have a separate sub-structure. I assume that's because we want the Data and ConfigRef for bootstrapping to be held together. Maybe we shouldn't worry so much about that, and just have BootstrapData and BootstrapRef as top-level fields?

Reply:
I also found this separate structure confusing at first. I don't see the rationale and as @dhellmann says, it is not consistent with other references. I second to simplify the model in this sense, unless the bootstrap structure has further motivation.

Reply:
The through process was around yaml/json UX. Option 1:
```yaml
spec:
  bootstrap:
    data: "xyz"
    configRef:
      apiVersion: foo/v1
      kind: Foo
      name: foo123
  ...
```

Option 2
```yaml
spec:
  bootstrapData: "xyz"
  bootstrapConfigRef:
      apiVersion: foo/v1
      kind: Foo
      name: foo123
  ...
```

With Option 1, we're grouping multiple things related to bootstrapping together. With Option 2, you have to write "bootstrap" multiple times.

If everyone is largely in agreement with the concept of a reference to a custom resource for bootstrap configuration, and a field to hold the bootstrap data, could we do the following?

1. Indicate that you agree w/the concept
1. Continue to debate the name and structure, **BUT**
1. After all other comments/issues/concerns are addressed, let the proposal merge in whatever state these bootstrap fields happen to be in, and we can continue to have the naming/structure discussion, if needed, after the proposal merges

Reply:
Per k8s [API Naming Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions):
> The name of a field referring to another resource of kind Foo by name should be called fooName. The name of a field referring to another resource of kind Foo by ObjectReference (or subset thereof) should be called fooRef.

Therefore I think it should be `bootstrapConfigRef` and option 2 (to avoid stuttering).

Reply:
I prefer option 1, better grouping of related fields/data. @davidewatson This kinda falls outside of that naming convention (although I understand why you make the point) because there is no `BootstrapConfig` kind here, but rather provider-specific CRDs, called arbitrary names.

Reply:
That's fair.